builder/llpcPipeline.h

```diff
@@ -167,6 +167,71 @@ enum class PrimitiveTopology : uint32_t
     PatchList = 10,
 };
 
+// Data format of vertex buffer entry. For ones that exist in GFX9 hardware, these match the hardware
+// encoding. But this also includes extra formats.
+enum BufDataFormat
+{
+    BufDataFormatInvalid                     = 0,
+    BufDataFormat8                           = 1,
+    BufDataFormat16                          = 2,
+    BufDataFormat8_8                         = 3,
+    BufDataFormat32                          = 4,
+    BufDataFormat16_16                       = 5,
+    BufDataFormat10_11_11                    = 6,
+    BufDataFormat11_11_10                    = 7,
+    BufDataFormat10_10_10_2                  = 8,
+    BufDataFormat2_10_10_10                  = 9,
+    BufDataFormat8_8_8_8                     = 10,
+    BufDataFormat32_32                       = 11,
+    BufDataFormat16_16_16_16                 = 12,
+    BufDataFormat32_32_32                    = 13,
+    BufDataFormat32_32_32_32                 = 14,
+    BufDataFormatReserved                    = 15,
+    // Extra formats not in GFX9 hardware encoding:
+    BufDataFormat8_8_8_8_BGRA,
+    BufDataFormat8_8_8,
+    BufDataFormat8_8_8_BGR,
+    BufDataFormat2_10_10_10_BGRA,
+    BufDataFormat64,
+    BufDataFormat64_64,
+    BufDataFormat64_64_64,
+    BufDataFormat64_64_64_64,
+};
+
+// Numeric format of vertex buffer entry. These match the GFX9 hardware encoding.
+enum BufNumFormat
+{
+    BufNumFormatUNORM                        = 0,
+    BufNumFormatSNORM                        = 1,
+    BufNumFormatUSCALED                      = 2,
+    BufNumFormatSSCALED                      = 3,
+    BufNumFormatUINT                         = 4,
+    BufNumFormatSINT                         = 5,
+    BufNumFormatSNORM_OGL                    = 6,
+    BufNumFormatFLOAT                        = 7,
+};
+
+// Rate of vertex input.
+enum
+{
+    VertexInputRateVertex = ~0,     // Vertex buffer has one element per vertex
```

Comment:
~0 is clearer as you don't need to understand how many Fs match the size of int you're using. :-)

Reply:
So I just want to understand why we assign a -1 to VertexInputRateVertex? Theoretically , VK_VERTEX_INPUT_RATE_VERTEX is a valid enumerant for us. It is reasonable we have: VertexInputRateNone = 0, VertexInputRateVertex = 1, VertexInputRateInstance  = 2.
Also, I see VertexInputRateNone is only used in a comment. What is the purpose of making it?

Reply:
(Unresolving this as I missed your comment after I marked it resolved.)

Sorry, I misunderstood your original question.

See the comment
// Other value N means vertex buffer has one element per N instances

So VertexInputRateInstance=1 is really a special case of the above, where no divide is needed because N=1. Then we need to assign VertexInputRateVertex and VertexInputRateNone to two values that cannot be used as the divisor N in the above. 0 and ~0 seem like good choices.

This scheme allows the rate and the instance divisor to be combined into a single value, which is neater than Vulkan does it. I guess Vulkan did not have the option to do it that way because having a divisor is an extension and it needed to maintain compatibility.

VertexInputRateNone is used in code, in VertexFetch::Run.

Reply:
Thanks. I think you can add part of the explanations as comment notes. So readers can understand your design purpose better :)

Reply:
Good plan; done.